build duckdb shell error with  “DISABLE_UNITY=1” for link stage issue

### What happens?

when build duckdb with "DISABLE_UNITY=1" . it report link error：
[ 90%] Linking CXX executable ../../duckdb
/usr/bin/ld: ../../extension/parquet/libparquet_extension.a(parquet_extension.cpp.o): in function `duckdb::ParquetExtension::Load(duckdb::DuckDB&)':
/home/zhaozhou/duckdb/origin/extension/parquet/parquet_extension.cpp:936: undefined reference to `duckdb::ExtensionUtil::RegisterFunction(duckdb::DatabaseInstance&, duckdb::TableFunctionSet)'
/usr/bin/ld: /home/zhaozhou/duckdb/origin/extension/parquet/parquet_extension.cpp:938: undefined reference to `duckdb::ExtensionUtil::RegisterFunction(duckdb::DatabaseInstance&, duckdb::TableFunctionSet)'
/usr/bin/ld: /home/zhaozhou/duckdb/origin/extension/parquet/parquet_extension.cpp:942: undefined reference to `duckdb::ExtensionUtil::RegisterFunction(duckdb::DatabaseInstance&, duckdb::TableFunctionSet)'
/usr/bin/ld: /home/zhaozhou/duckdb/origin/extension/parquet/parquet_extension.cpp:946: undefined reference to `duckdb::ExtensionUtil::RegisterFunction(duckdb::DatabaseInstance&, duckdb::TableFunctionSet)'
/usr/bin/ld: /home/zhaozhou/duckdb/origin/extension/parquet/parquet_extension.cpp:965: undefined reference to `duckdb::ExtensionUtil::RegisterFunction(duckdb::DatabaseInstance&, duckdb::CopyFunction)'
collect2: error: ld returned 1 exit status
make[2]: *** [tools/shell/CMakeFiles/shell.dir/build.make:127: duckdb] Error 1
make[1]: *** [CMakeFiles/Makefile2:6744: tools/shell/CMakeFiles/shell.dir/all] Error 2
make: *** [Makefile:136: all] Error 2


### To Reproduce

DISABLE_UNITY=1 make reldebug

### OS:

linux

### DuckDB Version:

latest main branch

### DuckDB Client:

latest main branch

### Full Name:

zhao zhou

### Affiliation:

clickhouse

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -679,8 +679,16 @@ function(build_loadable_extension NAME PARAMETERS)
   set(FILES ${ARGV})
   list(REMOVE_AT FILES 0)
   list(REMOVE_AT FILES 0)
-  target_link_libraries(${NAME}_extension duckdb_static)
+
   build_loadable_extension_directory(${NAME} "extension/${NAME}" "${PARAMETERS}" ${FILES})
+endfunction()
+
+function(build_static_extension NAME PARAMETERS)
+  # all parameters after name
+  set(FILES ${ARGV})
+  list(REMOVE_AT FILES 0)
+  add_library(${NAME}_extension STATIC ${FILES})
+  target_link_libraries(${NAME}_extension duckdb_static)
 endfunction()
 
 # Internal extension register function

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -679,7 +679,7 @@ function(build_loadable_extension NAME PARAMETERS)
   set(FILES ${ARGV})
   list(REMOVE_AT FILES 0)
   list(REMOVE_AT FILES 0)
-
+  target_link_libraries(${NAME}_extension duckdb_static)
   build_loadable_extension_directory(${NAME} "extension/${NAME}" "${PARAMETERS}" ${FILES})
 endfunction()
 

--- a/extension/autocomplete/CMakeLists.txt
+++ b/extension/autocomplete/CMakeLists.txt
@@ -6,7 +6,7 @@ include_directories(include)
 
 set(AUTOCOMPLETE_EXTENSION_FILES autocomplete_extension.cpp)
 
-add_library(autocomplete_extension STATIC ${AUTOCOMPLETE_EXTENSION_FILES})
+build_static_extension(autocomplete ${AUTOCOMPLETE_EXTENSION_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(autocomplete ${PARAMETERS}
                          ${AUTOCOMPLETE_EXTENSION_FILES})

--- a/extension/excel/CMakeLists.txt
+++ b/extension/excel/CMakeLists.txt
@@ -6,7 +6,6 @@ include_directories(numformat/include)
 include_directories(include)
 add_subdirectory(numformat)
 
-
 build_static_extension(excel excel_extension.cpp ${NUMFORMAT_OBJECT_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(excel ${PARAMETERS} excel_extension.cpp

--- a/extension/excel/CMakeLists.txt
+++ b/extension/excel/CMakeLists.txt
@@ -7,8 +7,7 @@ include_directories(include)
 add_subdirectory(numformat)
 
 
-build_static_extension(excel excel_extension.cpp
-                       ${NUMFORMAT_OBJECT_FILES})
+build_static_extension(excel excel_extension.cpp ${NUMFORMAT_OBJECT_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(excel ${PARAMETERS} excel_extension.cpp
                          ${NUMFORMAT_OBJECT_FILES})

--- a/extension/excel/CMakeLists.txt
+++ b/extension/excel/CMakeLists.txt
@@ -6,8 +6,9 @@ include_directories(numformat/include)
 include_directories(include)
 add_subdirectory(numformat)
 
-add_library(excel_extension STATIC excel_extension.cpp
-                                   ${NUMFORMAT_OBJECT_FILES})
+
+build_static_extension(excel excel_extension.cpp
+                       ${NUMFORMAT_OBJECT_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(excel ${PARAMETERS} excel_extension.cpp
                          ${NUMFORMAT_OBJECT_FILES})

--- a/extension/fts/CMakeLists.txt
+++ b/extension/fts/CMakeLists.txt
@@ -40,7 +40,7 @@ set(FTS_SOURCES
     ../../third_party/snowball/src_c/stem_UTF_8_tamil.cpp
     ../../third_party/snowball/src_c/stem_UTF_8_turkish.cpp)
 
-add_library(fts_extension STATIC ${FTS_SOURCES})
+build_static_extension(fts ${FTS_SOURCES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(fts ${PARAMETERS} ${FTS_SOURCES})
 

--- a/extension/httpfs/CMakeLists.txt
+++ b/extension/httpfs/CMakeLists.txt
@@ -6,7 +6,8 @@ add_extension_definitions()
 
 include_directories(include ../../third_party/httplib ../parquet/include)
 
-build_static_extension(httpfs s3fs.cpp httpfs.cpp crypto.cpp httpfs_extension.cpp)
+build_static_extension(httpfs s3fs.cpp httpfs.cpp crypto.cpp
+                       httpfs_extension.cpp)
 set(PARAMETERS "-warnings")
 build_loadable_extension(httpfs ${PARAMETERS} s3fs.cpp httpfs.cpp crypto.cpp
                          httpfs_extension.cpp)

--- a/extension/httpfs/CMakeLists.txt
+++ b/extension/httpfs/CMakeLists.txt
@@ -6,8 +6,7 @@ add_extension_definitions()
 
 include_directories(include ../../third_party/httplib ../parquet/include)
 
-add_library(httpfs_extension STATIC s3fs.cpp httpfs.cpp crypto.cpp
-                                    httpfs_extension.cpp)
+build_static_extension(httpfs s3fs.cpp httpfs.cpp crypto.cpp httpfs_extension.cpp)
 set(PARAMETERS "-warnings")
 build_loadable_extension(httpfs ${PARAMETERS} s3fs.cpp httpfs.cpp crypto.cpp
                          httpfs_extension.cpp)

--- a/extension/icu/CMakeLists.txt
+++ b/extension/icu/CMakeLists.txt
@@ -23,7 +23,7 @@ set(ICU_EXTENSION_FILES
     icu-timebucket.cpp
     icu-timezone.cpp)
 
-add_library(icu_extension STATIC ${ICU_EXTENSION_FILES})
+build_static_extension(icu ${ICU_EXTENSION_FILES})
 link_threads(icu_extension)
 disable_target_warnings(icu_extension)
 set(PARAMETERS "-no-warnings")

--- a/extension/jemalloc/CMakeLists.txt
+++ b/extension/jemalloc/CMakeLists.txt
@@ -9,8 +9,7 @@ add_subdirectory(jemalloc)
 
 set(JEMALLOC_EXTENSION_FILES jemalloc_extension.cpp ${JEMALLOC_OBJECT_FILES})
 
-add_library(jemalloc_extension STATIC ${JEMALLOC_EXTENSION_FILES})
-
+build_static_extension(jemalloc ${JEMALLOC_EXTENSION_FILES})
 # we do not do build_loadable_extension here because jemalloc is static-only
 
 install(

--- a/extension/json/CMakeLists.txt
+++ b/extension/json/CMakeLists.txt
@@ -31,7 +31,7 @@ set(JSON_EXTENSION_FILES
     json_functions/read_json_objects.cpp
     ${YYJSON_OBJECT_FILES})
 
-add_library(json_extension STATIC ${JSON_EXTENSION_FILES})
+build_static_extension(json ${JSON_EXTENSION_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(json ${PARAMETERS} ${JSON_EXTENSION_FILES})
 

--- a/extension/parquet/CMakeLists.txt
+++ b/extension/parquet/CMakeLists.txt
@@ -51,7 +51,7 @@ if(NOT CLANG_TIDY)
       ../../third_party/zstd/compress/zstd_opt.cpp)
 endif()
 
-add_library(parquet_extension STATIC ${PARQUET_EXTENSION_FILES})
+build_static_extension(parquet ${PARQUET_EXTENSION_FILES})
 
 set(PARAMETERS "-warnings")
 build_loadable_extension(parquet ${PARAMETERS} ${PARQUET_EXTENSION_FILES})

--- a/extension/parquet/CMakeLists.txt
+++ b/extension/parquet/CMakeLists.txt
@@ -52,7 +52,7 @@ if(NOT CLANG_TIDY)
 endif()
 
 add_library(parquet_extension STATIC ${PARQUET_EXTENSION_FILES})
-target_link_libraries(parquet_extension duckdb_static)
+
 set(PARAMETERS "-warnings")
 build_loadable_extension(parquet ${PARAMETERS} ${PARQUET_EXTENSION_FILES})
 

--- a/extension/parquet/CMakeLists.txt
+++ b/extension/parquet/CMakeLists.txt
@@ -52,6 +52,7 @@ if(NOT CLANG_TIDY)
 endif()
 
 add_library(parquet_extension STATIC ${PARQUET_EXTENSION_FILES})
+target_link_libraries(parquet_extension duckdb_static)
 set(PARAMETERS "-warnings")
 build_loadable_extension(parquet ${PARAMETERS} ${PARQUET_EXTENSION_FILES})
 

--- a/extension/sqlsmith/CMakeLists.txt
+++ b/extension/sqlsmith/CMakeLists.txt
@@ -10,7 +10,7 @@ set(SQLSMITH_SOURCES
     sqlsmith_extension.cpp statement_generator.cpp statement_simplifier.cpp
     fuzzyduck.cpp ${SQLSMITH_OBJECT_FILES})
 
-add_library(sqlsmith_extension STATIC ${SQLSMITH_SOURCES})
+build_static_extension(sqlsmith ${SQLSMITH_SOURCES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(sqlsmith ${PARAMETERS} ${SQLSMITH_SOURCES})
 

--- a/extension/tpcds/CMakeLists.txt
+++ b/extension/tpcds/CMakeLists.txt
@@ -6,7 +6,7 @@ include_directories(include)
 include_directories(dsdgen/include)
 add_subdirectory(dsdgen)
 
-add_library(tpcds_extension STATIC tpcds_extension.cpp ${DSDGEN_OBJECT_FILES})
+build_static_extension(tpcds tpcds_extension.cpp ${DSDGEN_OBJECT_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(tpcds ${PARAMETERS} tpcds_extension.cpp
                          ${DSDGEN_OBJECT_FILES})

--- a/extension/tpch/CMakeLists.txt
+++ b/extension/tpch/CMakeLists.txt
@@ -6,8 +6,7 @@ include_directories(dbgen/include)
 include_directories(include)
 add_subdirectory(dbgen)
 
-add_library(tpch_extension STATIC tpch_extension.cpp ${DBGEN_OBJECT_FILES})
-
+build_static_extension(tpch tpch_extension.cpp ${DBGEN_OBJECT_FILES})
 set(PARAMETERS "-warnings")
 build_loadable_extension(tpch ${PARAMETERS} tpch_extension.cpp
                          ${DBGEN_OBJECT_FILES})

--- a/extension/visualizer/CMakeLists.txt
+++ b/extension/visualizer/CMakeLists.txt
@@ -4,7 +4,7 @@ project(VisualizerExtension)
 
 include_directories(include)
 
-add_library(visualizer_extension STATIC visualizer_extension.cpp)
+build_static_extension(visualizer  visualizer_extension.cpp)
 set(PARAMETERS "-warnings")
 build_loadable_extension(visualizer ${PARAMETERS} visualizer_extension.cpp)
 install(

--- a/extension/visualizer/CMakeLists.txt
+++ b/extension/visualizer/CMakeLists.txt
@@ -4,7 +4,7 @@ project(VisualizerExtension)
 
 include_directories(include)
 
-build_static_extension(visualizer  visualizer_extension.cpp)
+build_static_extension(visualizer visualizer_extension.cpp)
 set(PARAMETERS "-warnings")
 build_loadable_extension(visualizer ${PARAMETERS} visualizer_extension.cpp)
 install(

--- a/src/include/duckdb/common/opener_file_system.hpp
+++ b/src/include/duckdb/common/opener_file_system.hpp
@@ -99,11 +99,11 @@ public:
 	bool IsPipe(const string &filename) override {
 		return GetFileSystem().IsPipe(filename);
 	}
-	virtual void RemoveFile(const string &filename) override {
+	void RemoveFile(const string &filename) override {
 		GetFileSystem().RemoveFile(filename);
 	}
 
-	virtual vector<string> Glob(const string &path, FileOpener *opener = nullptr) override {
+	vector<string> Glob(const string &path, FileOpener *opener = nullptr) override {
 		if (opener) {
 			throw InternalException("OpenerFileSystem cannot take an opener - the opener is pushed automatically");
 		}

